### ISSUE TITLE
Modifying Pretrained SCNN Model

### ISSUE THREAD

**Issue by virajdeokar** [XingangPan/SCNN#46]
I know there was a post about how to build an SCNN model to train on specific image sizes, but I was curious if it would be possible to take the pretrained SCNN model and change layers to allow for different image sizes. I would like to just fine tune the pretrained SCNN model to work on my dataset.

**Comment by cardwing** [comment]
In general, the probability map generation branch does not need modification since it is an encoder-decoder structure and can already recover original image size. As to the lane existence prediction branch, you just need to modify some layers (e.g., nn.View, nn.Linear) since the size of feature maps will be different if the size of input images changes.

**Comment by virajdeokar** [comment]
@cardwing Thanks for the response. Do you know exactly how I would go about modifying the pretrained model layers? I know there is a previous post about setting up a new SCNN model with different layer sizes, but this post doesn't provide enough information to extrapolate to doing the same on the pretrained SCNN model. I'm new to Torch/Lua and ML. 

Also, what would happen if I trained the model directly on images that are 1920x1080 instead of the 1600x576 size that they used? It seems the images get resized to the correct input size for SCNN. Would doing this negatively impact my results since the images I am using have a different aspect ratio than the images that they used? Thanks again for your help.

**Comment by virajdeokar** [comment]
Like in https://github.com/XingangPan/SCNN/issues/7 , they mention changing the Num and Length values in SCNN-gen. Since we are not generating a new SCNN model and rather altering a prexisitng and pretrained one, how would we accomplish this?

**Comment by cardwing** [comment]
@virajdeokar , a naive solution is to resize new images to 1600 x 590. However, this may harm the performance as you have pointed out above. A smarter solution is to not modify the image size but instead modify the model. As mentioned above, you just need to modify the lane existence prediction branch. More specifically, you just need to load the original model via `model = torch.load('you_model_path'):cuda()` and `model = model:get(1)` . Then remove nn.View and nn.Linear via `model:get(x):get(x)...:remove(x)`. Eventually, you can print the architecture of the model via `print('Model:\n' .. model:__tostring())`

**Comment by virajdeokar** [comment]
@cardwing Ok. That makes sense. I can easily change the classification layer of the pretrained model, but I am concerned about some of the earlier layers that are generate in the SCNN model. If you look at the SCNN-gen.lua file, there is this function:
`function buildPass(d,width,dim,scale,kw) --d=1:down-up d=2:right-left
   local pass = nn.Sequential()
   local length = 0
   if d==1 then
      Num = 36/scale
      length = 100/scale
   else
      Num = 100/scale
      length = 36/scale
   end
   local num = Num/width
   local function buildParal()
      local seq = nn.Sequential()
      seq:add(nn.SplitTable(d+1, 3)) --128 36
      -- view fom 100 to 1,1,100
      local paralView = nn.ParallelTable()
      for i=1,Num do
         local view = nn.Sequential()
         view:add(nn.Contiguous())
         if d==1 then
            view:add(nn.View(dim, 1, length):setNumInputDims(2))
         else
            view:add(nn.View(dim, length, 1):setNumInputDims(2))
         end
         paralView:add(view)
      end
      seq:add(paralView)
...(continues)`

The Num and length variables of 100 and 36 are the input/output image width and height divided by 8. These variables are essential to when SCNN generates nn.View's in its earlier layers. I am worried that these layers will cause problems if we only changed the classification layer of SCNN since these layers aren't adjusted for the new image size. What do you think? If I had to change these layers, I think it would be rather hard. Would there be a way to generate a new SCNN model using SCNN-gen.lua and then transfer the weights of the pretrained model?

**Comment by cardwing** [comment]
@virajdeokar , since your new image size is 1920 x 1080, which is divisible by 8, you can modify the message passing part as well as the classification branch. If you really need to modify the current version of SCNN to accommodate new images (1920 x 1080), I can write one version to you.

**Comment by virajdeokar** [comment]
@cardwing That would be amazing if you could do that! I think it would be very helpful for other people as well. Yes, my input image size is 1920x1080. However, I wanted my classification image size to be 768x432 (this is 1920x1080 divided by 2.5). This fits the requirement that the classification height/width by divisible by 16. Thus, my new View size would be 6,480 (5x768x432/16/16) for the classification branch. In the buildPass function above the 100 would thus be replaced with 96 and the 36 with 54.

**Comment by cardwing** [comment]
@virajdeokar , I will try to write one version and you can check if it satisfies your requirements.

**Comment by virajdeokar** [comment]
@cardwing Great! Thanks again.

**Comment by cardwing** [comment]
@virajdeokar , remember to change the file name from .txt to .lua. Besides, use `CUDA_VISIBLE_DEVICES="0" th SCNN-gen.lua` to run the file. I have tested the model and it should work (testing code is also in this file). Good luck!

**Comment by cardwing** [comment]
After run the file, you should see the model structure and the output size of the model. Like this:
   2
   5
 432
 768
[torch.LongStorage of size 4]

 2
 4
[torch.LongStorage of size 2]

**Comment by cardwing** [comment]
@virajdeokar , if the modified model can work, you can close this issue now ^_^!

**Comment by virajdeokar** [comment]
@cardwing Thanks. This file will generate a new SCNN model as far as I can tell. My question is, would it be possible to load weights from the pretrained SCNN model called 
vgg_SCNN_DULR_w9.t7 that they provide for download on the GitHub repo? My goal is fine tune SCNN but using SCNN-gen.lua creates a new SCNN model which will require extensive training as far as I can tell?

**Comment by virajdeokar** [comment]
If this is not possible/ you do not believe that it will improve results greatly, please let me know.

**Comment by cardwing** [comment]
@virajdeokar , If you use a different image size, an ImageNet pre-trained model should be enough (vgg.t7). However, if you really want to load weights from a pre-trained SCNN model, you can just modify the model loading path in Line 5 and comment Line 6 and Line 140. It can already meet your requirements.

**Comment by cardwing** [comment]
If you meet a data parallel problem, just use `model=model:get(1)`.

**Comment by virajdeokar** [comment]
@cardwing Okay. That would only change the classification layer of the pre-trained SCNN model. Would it be possible to just generate a new SCNN model like we do now, but have the weights from the pre-trained SCNN model be loaded into our new SCNN model? This way, we could have an SCNN model that matches our image size but also doesn't have random weights to start off with. so training would be faster? I'm looking at line 58 in the file. Correct me if random weights vs pre-trained weights won't make much of a difference in training since image sizes are different. Thanks again.

**Comment by cardwing** [comment]
[SCNN-gen-latest.txt](https://github.com/XingangPan/SCNN/files/2261388/SCNN-gen-latest.txt)

**Comment by cardwing** [comment]
@virajdeokar , just have a look at the SCNN-gen-latest.txt. It should meet your requirements. Besides, using a SCNN pre-trained model can bring more performance gains since it is equivalent to using more training data.

**Comment by cardwing** [comment]
However, it is just my understanding. You can test both training strategies and check which one is better since deep learning is empirical.
